csaf/src/metadata_retriever/mod.rs

```diff
@@ -0,0 +1,209 @@
+            .fetch::<Option<Json<ProviderMetadata>>>(provider_metadata_url.clone())
+            .await?;
+        Ok(metadata)
+    }
+
+    /// Retrieve provider metadata through the full well known URL.
+    pub async fn retrieve_metadata_by_well_known_url(&self) -> Result<DetectionType, Error>
+    where
+        Self: Sized + Send,
+    {
+        log::trace!("Starting retrieve provider metadata from full provided discovery URL");
+        let provider_metadata =
+            Self::fetch_metadata_from_url(&self.fetcher, self.provider_metadata_url.clone())
+                .await?;
+
+        if let Some(provider_metadata) = provider_metadata {
+            Ok(DetectionType::WellKnowPath(provider_metadata.into_inner()))
+        } else {
+            Err(Error::EmptyProviderMetadata(
+                self.provider_metadata_url.clone().to_string(),
+            ))
+        }
+    }
+
+    /// Retrieve provider metadata through the DNS path of provided URL.
+    pub async fn retrieve_metadata_by_dns(&self) -> Result<DetectionType, Error>
+    where
+        Self: Sized + Send,
+    {
+        log::trace!("Starting retrieve provider_metadata from DNS path of provided discovery URL ");
+        if let Some(host_url) = self.provider_metadata_url.clone().host_str() {
+            log::warn!("Querying provider metadata url DNS  {:?}", host_url);
+            let host_url_string = host_url.replace("www.", "");
+            let host_url_string = format!("https://csaf.data.security.{}", &host_url_string);
+            let dns_path = Url::parse(host_url_string.as_str())?;
+
+            let provider_metadata =
+                Self::fetch_metadata_from_url(&self.fetcher, dns_path.clone()).await?;
+
+            if let Some(provider_metadata) = provider_metadata {
+                Ok(DetectionType::WellKnowPath(provider_metadata.into_inner()))
+            } else {
+                Err(Error::EmptyProviderMetadata(dns_path.to_string()))
+            }
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+
+    /// Retrieve provider metadata through the security text of provided URL.
+    pub async fn retrieve_metadata_by_security_text(
+        &self,
+        security_txt_path: &str,
+    ) -> Result<DetectionType, Error>
+    where
+        Self: Sized + Send,
+    {
+        log::trace!(
+            "Starting retrieve provider_metadata from security text of provided discovery URL "
+        );
+        if let Some(host_url) = self.provider_metadata_url.clone().host_str() {
+            let host_url_string = format!("https://{}", &host_url);
+            let host_url = Url::parse(host_url_string.as_str())?;
+            log::warn!(
+                "Querying provider metadata url from security text of {:?}",
+                host_url_string
+            );
+            let provider_metadata_path =
+                Self::get_url_from_security_text(&self.fetcher, host_url.join(security_txt_path)?)
```

Comment:
In case we have an entry in the security.txt file, the situation of a 404 is different! If the provider file referenced from that security.txt is missing, that would indeed be an error.